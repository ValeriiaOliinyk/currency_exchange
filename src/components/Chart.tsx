import React from "react";
import { Bar } from "react-chartjs-2";

import { chart } from "../helpers/chartData";
import { Container, Title } from "../styles/components";

export const Chart = () => {
  return (
    <>
      <Container>
        <Title>The top 10 most traded currencies in the world</Title>
        <Bar
          data={chart.data}
          width={50}
          height={100}
          options={{
            legend: {
              display: true,
              position: "bottom"
            }
          }}
        />
      </Container>
    </>
  );
};
