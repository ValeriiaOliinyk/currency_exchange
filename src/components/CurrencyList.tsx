import React from "react";
import PropTypes from "prop-types";
import { useSelector, useDispatch } from "react-redux";

import {
  addFavorite,
  deleteFavorite,
  getFavorites,
  getRegularCurrency,
  getNumberOfFavorites,
} from "../redux/ducks/currency";

// Components
import { MainLoader, Empty } from ".";
import {
  ListFavorite,
  BtnFavorite,
  BtnEmpty,
  ItemFavorites,
  ItemContainer,
  Value,
} from "../styles/components";

export function CurrencyList() {
  const dispatch = useDispatch();
  const favorites: Array<string> | undefined = useSelector(getFavorites);
  const regularCurrency: Array<string> | undefined = useSelector(
    getRegularCurrency
  );
  const numberOfFavorites: Number | undefined = useSelector(
    getNumberOfFavorites
  );

  const toggleFavAction = (text: string) => {
    if (!favorites.includes(text)) {
      dispatch(addFavorite(text));
    }

    if (favorites.includes(text)) {
      dispatch(deleteFavorite(text));
    }
  };

  return (
    <>
      <ListFavorite favorite>
        {numberOfFavorites > 0 ? (
          favorites.map((item) => (
            <ItemFavorites key={item}>
              <ItemContainer>
                <Value>{item}</Value>
                <BtnFavorite
                  className="Btn"
                  type="button"
                  onClick={() => toggleFavAction(item)}
                ></BtnFavorite>
              </ItemContainer>
            </ItemFavorites>
          ))
        ) : (
          <Empty />
        )}
      </ListFavorite>
      <ListFavorite>
        {regularCurrency ? (
          regularCurrency.map((item) => (
            <ItemFavorites key={item}>
              <ItemContainer>
                <Value>{item}</Value>
                <BtnEmpty
                  type="button"
                  onClick={() => toggleFavAction(item)}
                ></BtnEmpty>
              </ItemContainer>
            </ItemFavorites>
          ))
        ) : (
          <MainLoader />
        )}
      </ListFavorite>
    </>
  );
}

CurrencyList.defaultProps = {
  numberOfFavorites: 0,
  favorites: [],
};

CurrencyList.propTypes = {
  numberOfFavorites: PropTypes.number,
  favorites: PropTypes.arrayOf(PropTypes.string),
  regularCurrency: PropTypes.arrayOf(PropTypes.string),
};
