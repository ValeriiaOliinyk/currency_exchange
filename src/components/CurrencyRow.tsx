import React, { useState, useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import PropTypes from "prop-types";
import { Formik } from "formik";
import {
  loadData,
  getCurrencyArray,
  getFromCurrency,
  getToCurrency,
  getExchangeRate,
  updateData,
  getUpdatedData,
  addExchangeRate,
} from "../redux/ducks/currency";

import { validationSchema } from "../helpers/validation";
import { InitialValues, CurrencyTypes } from "../helpers/interfaces";

// Components
import {
  FormBox,
  Option,
  FormEqually,
  FormCurrency,
  FormControl,
  FormSelect,
  ErrorMessage,
} from "../styles/components";

// Styles
import { Container } from "react-bootstrap";

export function CurrencyRow() {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(loadData());
  }, [dispatch]);

  const currency: Array<string> | undefined = useSelector(getCurrencyArray);
  const selectedFromCurrency: string | undefined = useSelector(getFromCurrency);
  const selectedToCurrency: string | undefined = useSelector(getToCurrency);
  const selectedExchangeRate: number | undefined = useSelector(getExchangeRate);
  const updatedData: CurrencyTypes = useSelector(getUpdatedData);

  const [fromCurrency, setFromCurrency] = useState<string | undefined>(
    selectedFromCurrency
  );
  const [toCurrency, setToCurrency] = useState<string | undefined>(
    selectedToCurrency
  );
  const [exchangeRate, setExchangeRate] = useState<number>(
    selectedExchangeRate
  );
  const [amount, setAmount] = useState<number>(1);
  const [amountInFromCurrency, setAmountInFromCurrency] = useState<boolean>(
    true
  );

  const firstExchange = useRef<number>(exchangeRate);

  useEffect(() => {
    setToCurrency(selectedToCurrency);
  }, [selectedToCurrency]);

  useEffect(() => {
    if (updatedData.rates && toCurrency) {
      setExchangeRate(updatedData.rates[toCurrency]);
    }
  }, [dispatch, toCurrency, updatedData.rates]);

  useEffect(() => {
    dispatch(addExchangeRate(firstExchange.current));
  }, [dispatch]);

  useEffect(() => {
    if (fromCurrency && toCurrency) {
      dispatch(updateData(fromCurrency, toCurrency));
    }
  }, [dispatch, fromCurrency, toCurrency]);

  const getInitialValues = (): InitialValues => {
    return {
      fromAmount: amountInFromCurrency
        ? amount
        : (amount / exchangeRate).toFixed(1),
      toAmount: amountInFromCurrency
        ? (amount * exchangeRate).toFixed(1)
        : amount,
    };
  };

  return (
    <Container>
      <Formik
        enableReinitialize={true}
        initialValues={getInitialValues()}
        validationSchema={validationSchema}
        onSubmit={() => {}}
      >
        {({ values, errors, touched }) => (
          <FormCurrency>
            <FormBox>
              <FormControl
                type="number"
                name="fromAmount"
                value={values.fromAmount}
                onChange={(e: {
                  target: { value: React.SetStateAction<number> };
                }) => {
                  setAmount(e.target.value);
                  setAmountInFromCurrency(true);
                }}
              />
              <FormSelect
                as="select"
                value={fromCurrency}
                onChange={(e: {
                  target: { value: React.SetStateAction<string | undefined> };
                }) => setFromCurrency(e.target.value)}
              >
                {currency &&
                  currency.map((item) => (
                    <Option key={item} value={item}>
                      {item}
                    </Option>
                  ))}
              </FormSelect>
            </FormBox>
            {errors.fromAmount && touched.fromAmount ? (
              <ErrorMessage>{errors.fromAmount}</ErrorMessage>
            ) : (
              <ErrorMessage transparent>Transparent</ErrorMessage>
            )}
            <FormEqually />
            <FormBox>
              <FormControl
                type="number"
                name="toAmount"
                value={values.toAmount}
                onChange={(e: {
                  target: { value: React.SetStateAction<number> };
                }) => {
                  setAmount(e.target.value);
                  setAmountInFromCurrency(false);
                }}
              />
              <FormSelect
                as="select"
                value={toCurrency}
                onChange={(e: {
                  target: { value: React.SetStateAction<string | undefined> };
                }) => setToCurrency(e.target.value)}
              >
                {currency &&
                  currency.map((item) => (
                    <Option key={item} value={item}>
                      {item}
                    </Option>
                  ))}
              </FormSelect>
            </FormBox>
            {errors.toAmount && touched.toAmount ? (
              <ErrorMessage>{errors.toAmount}</ErrorMessage>
            ) : (
              <ErrorMessage transparent>Transparent</ErrorMessage>
            )}
          </FormCurrency>
        )}
      </Formik>
    </Container>
  );
}

CurrencyRow.defaultProps = {
  currency: [],
  selectedFromCurrency: "",
  selectedToCurrency: "",
  selectedExchangeRate: 0,
};

CurrencyRow.propTypes = {
  currency: PropTypes.arrayOf(PropTypes.string),
  selectedFromCurrency: PropTypes.string,
  selectedToCurrency: PropTypes.string,
  selectedExchangeRate: PropTypes.number,
};
