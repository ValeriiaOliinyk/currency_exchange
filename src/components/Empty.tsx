import React from "react";

import { EmptyFavorites } from "./styled";

export const Empty = () => (
  <EmptyFavorites>No favorite currencies</EmptyFavorites>
);
