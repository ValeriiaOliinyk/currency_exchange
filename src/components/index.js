export { Convert } from "./Convert/Convert";
export { CurrencyList } from "./CurrencyList/CurrencyList";
export { CurrencyRow } from "./CurrencyRow/CurrencyRow";
export { Empty } from "./Empty/Empty";
export { MainLoader } from "./MainLoader/MainLoader";
export { Navigation } from "./Navigation/Navigation";
