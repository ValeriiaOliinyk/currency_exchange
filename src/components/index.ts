export { Convert } from "./Convert";
export { CurrencyList } from "./CurrencyList";
export { CurrencyRow } from "./CurrencyRow";
export { Empty } from "./Empty";
export { MainLoader } from "./MainLoader";
export { Navigation } from "./Navigation";
export { Gallery } from "./Gallery";
export { Chart } from "./Chart";
