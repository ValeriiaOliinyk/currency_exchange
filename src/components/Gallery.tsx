import React from 'react';
import Grid from '@material-ui/core/Grid';

import { Image } from '../styles/components';

export const Gallery = () => (
  <Grid container spacing={2} justify="center">
    <Grid item>
      <Image src="./images/money-2696229_640.jpg" />
    </Grid>
    <Grid item>
      <Image src="./images/money-2696228_640.jpg" />
    </Grid>
    <Grid item>
      <Image src="./images/money-2696229_640.jpg" />
    </Grid>
    <Grid item>
      <Image src="./images/money-2696234_640.jpg" />
    </Grid>
    <Grid item>
      <Image src="./images/money-2724235_640.jpg" />
    </Grid>
    <Grid item>
      <Image src="./images/money-2724241_640.jpg" />
    </Grid>
  </Grid>
);
