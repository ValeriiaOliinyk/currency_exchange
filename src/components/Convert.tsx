import React from 'react';

// Components
import { CurrencyRow } from './CurrencyRow';
import { Title, Container } from '../styles/components';

export function Convert() {
  return (
    <Container>
      <Title>Convert</Title>
      <CurrencyRow />
    </Container>
  );
}
