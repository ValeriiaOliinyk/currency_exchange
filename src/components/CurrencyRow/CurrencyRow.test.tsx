import React from "react";
import * as reactRedux from "react-redux";
import { shallow, ShallowWrapper } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { currencyResponse } from "../../helpers/fakeResponse";
import { CurrencyRow } from "../../components";

const initialState = {
  currency: currencyResponse,
  rates: { from: "EUR", to: "RUB" },
  rate: 1.558,
  updatedData: { base: "EUR", date: "2020-11-04", rates: { CAD: 1.5411 } },
};

describe("CurrencyRow component renders correctly", async () => {
  let tree: ShallowWrapper<any>;
  let store: any = configureStore()(initialState);
  beforeEach(() => {
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((state) => store.getState().getCurrency);
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((state) => store.getState().updatedData);
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((state) => store.getState().currency);
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((state) => store.getState().rates);

    jest
      .spyOn(reactRedux, "useDispatch")
      .mockImplementation(() => store.dispatch);

    tree = shallow(
      <Provider store={store}>
        <CurrencyRow />
      </Provider>
    )
      .dive()
      .dive();
  });

  test("Component renders correctly", async () => {
    expect(tree).toMatchSnapshot();
  });
});
