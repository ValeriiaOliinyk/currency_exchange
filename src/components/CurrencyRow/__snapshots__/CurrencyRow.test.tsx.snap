// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyRow component renders correctly Component renders correctly 1`] = `ShallowWrapper {}`;
